### ISSUE TITLE
Explore @lemire's rolling hash function as an alternative for Counttable

### ISSUE THREAD

**Issue by ctb** [dib-lab/khmer#1616]
See https://github.com/lemire/rollinghashcpp/blob/master/example.cpp

Should be much faster than murmurhash.

**Comment by betatim** [comment]
Once #1511 is merged this is going to be easy :)

As a first step I would take the code from lemire/rollinghashcpp, put what we need in `third-party/` and add a `KarpRabinKmerHashIterator`

**Comment by asl** [comment]
Note that the version above only supports 19 bit hashes. Something like https://github.com/bcgsc/ntHash is more universal, however, it's under GPLv3

**Comment by lemire** [comment]
@asl 

>Note that the version above only supports 19 bit hashes. 

 19 is the default, not the sole setting.

19 is an odd default, I'll grant you that.

**Comment by asl** [comment]
@lemire 

Probably I was confused by the following code:
```
if(wordsize == 19) {
            irreduciblepoly = 1 + (1<<1) + (1<<2) + (1<<5)
                               + (1<<19);
        } else if (wordsize == 9) {
            irreduciblepoly = 1+(1<<2)+(1<<3)+(1<<5)+(1<<9);
        } else {
            cerr << "unsupported wordsize "<<wordsize << " bits, try 19 or 9"<< endl;
        }
```

**Comment by lemire** [comment]
@asl Right. One of our templates has weird limitations,  but that's not the one you are thinking about using, I believe.

**Comment by sjackman** [comment]
@mohamadi has the fast rolling hash implementation ntHash https://github.com/bcgsc/ntHash
https://doi.org/10.1093/bioinformatics/btw397
that is used by his k-mer cardinality estimation tool ntCard https://github.com/bcgsc/ntCard
https://doi.org/10.1093/bioinformatics/btw832

**Comment by asl** [comment]
@sjackman Unfortunately, it's GPL3. And I already mentioned it above :) Anyway, #1699 contains standalone rolling hash implementation extracted from the code of @lemire

**Comment by sjackman** [comment]
Ah, woops. My mistake.
@mohamadi How do you feel about dual licensing ntHash GPL-3 and MIT?

**Comment by mohamadi** [comment]
@sjackman @asl Sure. I'll make it dual licensing now.

**Comment by sjackman** [comment]
Sounds like it's a moot point for khmer, since you already have #1699. If the license prevented khmer from adopting ntHash, it could be a similar impediment to other projects. Dual licensing could help its uptake.

**Comment by sjackman** [comment]
Awesome. Good show, Hamid.
I'd suggest changing the `LICENSE` file to the more permissive MIT license, and mention in the README.md that it's dual licensed MIT or GPL-3.

**Comment by sjackman** [comment]
@asl If the license is the only thing holding you back from using a project, it never hurts to ask the author to consider relicensing.

**Comment by sjackman** [comment]
On second thought, there's not much benefit to dual licensing. May as well just go MIT license only.

**Comment by mohamadi** [comment]
I just changed the license from GPL3 to only MIT.

**Comment by sjackman** [comment]
Thanks, Hamid!

**Comment by standage** [comment]
Great. Thanks for the responses!

**Comment by sjackman** [comment]
Happy to!

**Comment by ctb** [comment]
I think this is fixed by #1792!
